---
title: 'hackseq: Trends in Bioinformatics'
author: "Diana Lin and Emma Garlock"
date: "14/11/2019"
output:
  ioslides_presentation:
    logo: R-LadiesGlobal.png
---


```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width=12, fig.height=8)
# tidy.opts=list(width.cutoff=60),tidy=TRUE
# knitr::opts_chunk$set(echo = TRUE, error = FALSE, message = FALSE, warning = FALSE, fig.width=12, fig.height=8)
```
## `hackseq`
`hackseq` is a Vancouver-based hackathon focused on genomics. They want to bring individuals with diverse backgrounds together to collaborate on scientific questions and problems in genomics.

Their philosophy is __open-source__, __open-notebook__, __open science__.

```{r hackseq, out.width="500px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://github.com/dy-lin/hs19-trends/raw/master/workshop/hackseq.png")
```

<center><font size="3">Source: https://www.hackseq.com</font><center>

## Project Goals
__Bioinformatics:__ the development and use of computational methods in genetics and genomics

### __Graphic Visualization of trends in Bioinformatics__

1. Usage of bioinformatic tools and technqiues (ex. sequence alignment, genome assembly, metagenomics etc)
2. Relationship of tool development within analytical pipelines
3. Geographic hotspots for development of bioinformatic tools and techniques

__End product__: A visualization tool of trends in Bioinformatics, which can help prospective graduate students choose an institution or area of research.

## Trends in Bioinformatics

#### __Workflow__

1. Divide the field of Bioinformatics into topics
    + Sequencing, Assembly, ...
2. Determine search terms for each subjects
    + Sequencing: Sanger, next-generation sequencing, ...
3. Analyze trends by webscraping/textmining
    + using `fulltext`, `pubchunks`

##  Analogy
Field | Topic | Search Terms 
------|-------|----------------
Bioinformatics | Sequencing | Sanger, next-generation sequencing, ...
Bioinformatics | Assembly | Short read assembly, long read assembly, ...
Fruits | Apples | Ambrosia, Gala, McIntosh, Granny Smith, ...
Fruits | Oranges | Navel, Mandarin, Tangerine, Clementine, ...


```{r setup, include = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
library(gganimate)
library(knitr)
```

```{r load sequencing data, include=FALSE, echo = FALSE}
# Load data for all 7 sequencing types
url <- "https://raw.githubusercontent.com/dy-lin/hs19-trends/master/data/sequencing-specific.csv"
seq_spec <- read_csv(url)
```

## Webscraping Results

```{r overview code, echo = FALSE, message = FALSE}
resFound <- read_csv("https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/data/webscrape_overall.csv")

overview <- resFound %>%
  group_by(topic, db) %>%
  summarize(sum = sum(found)) %>%
  ggplot(aes(x = topic, y = sum)) +
  geom_col(aes(fill = db)) + 
  theme(axis.text.x = element_text(angle = 30),
        text = element_text(size = 20),
        axis.title = element_text(size =18),
        axis.text = element_text(size=12)
        ) +
  scale_y_continuous(label = scales::comma_format()) +
  labs(title = "", y = "# Papers Found", x = "Topic", fill = "Database") + 
  ggtitle("Overview of Topics Across Databases")
```
Dataframe:
```{r print overview table, echo = FALSE}
head(resFound, 12L)
```

## Webscraping Results {.smaller}

Stacked Bar Graph: 
```{r print overview graph, echo = FALSE, out.width = "700px"}
print(overview)
```

<center><font size="3">Source: https://dy-lin.github.io/hs19-trends/R/overview.html</font></center>
<!-- ```{r, out.width = "830px", echo = FALSE}
knitr::include_graphics("https://dy-lin.github.io/hs19-trends/figures/overview.png")
```-->

## Our Dataset {.smaller}
Let's look at the search terms of one specific topic, one specific jounral:

__Sequencing in the Public Library of Science (PLoS)__
```{r, echo = FALSE}
head(seq_spec, 10L)
```
To follow along with some live-coding, download this worksheet:
<center>http://tiny.cc/rladies-ws1</center><!--or
https://github.com/dy-lin/hs19-trends/blob/master/workshop/worksheet_dlin.Rmd-->

# _Racing_ Bar Graph 

<!--## What is a Racing Bar Graph?
```{r , out.width="800px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://i2.wp.com/michaeltoth.me/figures/city_populations/create_graph-1.gif?w=456&ssl=1")
```
<center><font size="3">Source: https://www.r-bloggers.com/how-to-create-a-bar-chart-race-in-r-mapping-united-states-city-population-1790-2010/</font></center>-->

## What is a Racing Bar Graph?
```{r , out.width="700px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://emilykuehler.github.io/img/barplot_race.gif")
```
<center><font size="3">Source: https://emilykuehler.github.io/bar-chart-race/</font></center>

## What is a Racing Bar Graph?
```{r , out.width="700px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://emilykuehler.github.io/img/nba_scoring_leaders.gif")
```
<center><font size="3">Source: https://emilykuehler.github.io/bar-chart-race/</font></center>

## Regular Bar Graph {.smaller}

First, let's load our dataset and make a _regular_ bar graph:
```{r load data, message = FALSE}
base <- "https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/"
seq_spec <- "data/sequencing-specific-processed.csv"
url <- paste0(base, seq_spec)
seq_data <- read_csv(url)
```

A glimpse into the dataset (top 5):
```{r glimpse, echo = FALSE}
seq_data %>% drop_na() %>% arrange(desc(total)) %>% head(5L)  %>% kable()
```

## Regular Bar Graph
```{r regular bar, out.width="600px"}
reg <- seq_data %>%
  drop_na() %>%
  ggplot(aes(x = topic, y = cum_total, fill = topic)) +
  geom_col() + 
  facet_wrap(~ Year, ncol = 8) + 
  coord_flip() +
  scale_fill_viridis_d() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "bottom") +
  geom_text(aes(y = cum_total, 
                label = topic), 
            hjust = "left", 
            fontface = "bold", 
            nudge_y = 50)
```

## Regular Bar Graph {.vcenter .flexbox}
```{r, out.width="800px", echo = FALSE}
print(reg)
```



## _Racing_ Bar Graph
In order to make a racing bar chart, where the bars overtake one another, we need to rank the topics for each year:

```{r ordering racing bars}
ordered_df <- NULL

for (yr in 2003:2019) {
  order <- seq_data %>% 
    filter(Year == yr) %>% 
    arrange(cum_total) %>% 
    mutate(ordering = row_number())
  
  ordered_df <- ordered_df %>% rbind(order)
}
```

## _Racing_ Bar Graph {.smaller}
Here's what the dataset looks like now:
```{r, echo = FALSE}
ordered_df %>% head(9L) %>% kable()
```

## _Racing_ Bar Graph {.smaller}
Here is an overview of the code:
```{r racing bars animate overall, eval = FALSE}
# plot
p <- ordered_df %>% 
  ggplot(aes(ordering, group = topic)) +
  geom_col(aes(y = cum_total, width = 0.9, fill = topic)) +
  geom_text(aes(y = cum_total, label = topic), 
            hjust = "left", fontface = "bold", nudge_y = 50) +
  coord_cartesian(clip = "off", expand = FALSE) +
  scale_fill_viridis_d() +
  coord_flip() +
# animate
  transition_states(Year, transition_length = 8, state_length = 4, wrap = FALSE) +
  ease_aes("cubic-in-out") +
# aesthetics
  labs(subtitle = "Trends in sequencing methods", title = "Year {closest_state}", 
       y = "cumulative total papers") +
  theme(plot.background = element_blank(), legend.position = "none",
        axis.ticks.y = element_blank(), axis.text.y = element_blank(),
        text = element_text(size=14), plot.title = element_text(size = 35)) +
  ylim(0,1300) +
  xlab("") 
```

## _Racing_ Bar Graph {.smaller}
Let's start plotting:
```{r racing bars animate plot, eval = FALSE}
# plot
### <b>
p <- ordered_df %>% 
  ggplot(aes(ordering, group = topic)) +
  geom_col(aes(y = cum_total, width = 0.9, fill = topic)) +
  geom_text(aes(y = cum_total, label = topic), 
            hjust = "left", fontface = "bold", nudge_y = 50) +
  coord_cartesian(clip = "off", expand = FALSE) +
  scale_fill_viridis_d() +
  coord_flip() +
### </b>
# animate
  transition_states(Year, transition_length = 8, state_length = 4, wrap = FALSE) +
  ease_aes("cubic-in-out") +
# aesthetics
  labs(subtitle = "Trends in sequencing methods", title = "Year {closest_state}", 
       y = "cumulative total papers") +
  theme(plot.background = element_blank(), legend.position = "none",
        axis.ticks.y = element_blank(), axis.text.y = element_blank(),
        text = element_text(size=14), plot.title = element_text(size = 35)) +
  ylim(0,1300) +
  xlab("") 
```
## _Racing_ Bar Graph {.smaller}
`coord_cartesian(clip = "off", expand = FALSE)`

#### Usage:

`coord_cartesian(xlim = NULL, ylim = NULL, expand = TRUE, default = FALSE, clip = "on")`

#### Description:
The Cartesian coordinate system is the most familiar, and common, type of coordinate system. Setting limits on the coordinate system will zoom the plot (like you're looking at it with a magnifying glass), and will not change the underlying data like setting limits on a scale will.

#### Arguments
* `expand`: If `TRUE`, the default, adds a small expansion factor to the limits to ensure that data and axes don't overlap. If `FALSE`, limits are taken exactly from the data or `xlim`/`ylim`.
* `clip`: Should drawing be clipped to the extent of the plot panel? A setting of "`on`" (the default) means yes, and a setting of "`off`" means no. In most cases, the default of "`on`" should not be changed, as setting clip = "`off`" can cause unexpected results. It allows drawing of data points anywhere on the plot, including in the plot margins. 

## _Racing_ Bar Graph {.smaller}
Next, to set up some animation parameters:
```{r racing bars animate animation, eval = FALSE}
# plot
p <- ordered_df %>% 
  ggplot(aes(ordering, group = topic)) +
  geom_col(aes(y = cum_total, width = 0.9, fill = topic)) +
  geom_text(aes(y = cum_total, label = topic), 
            hjust = "left", fontface = "bold", nudge_y = 50) +
  coord_cartesian(clip = "off", expand = FALSE) +
  scale_fill_viridis_d() +
  coord_flip() +
# animate
### <b>
  transition_states(Year, transition_length = 8, 
                    state_length = 4, wrap = FALSE) +
  ease_aes("cubic-in-out") +
### </b>
# aesthetics
  labs(subtitle = "Trends in sequencing methods", y = "cumulative total papers",
       title = "Year {closest_state}") +
  theme(plot.background = element_blank(), legend.position = "none",
        axis.ticks.y = element_blank(), axis.text.y = element_blank(),
        text = element_text(size=14), plot.title = element_text(size = 35)) +
  ylim(0,1300) +
  xlab("") 
```

## _Racing_ Bar Graph {.smaller}
`transition_states(Year, transition_length = 8, state_length = 4, wrap = FALSE)`

#### Usage: 

`transition_states(states, transition_length = 1, state_length = 1, wrap = TRUE)`

#### Description:
This transition splits your data into multiple states based on the levels in a given column, much like `ggplot2::facet_wrap()` splits up the data in multiple panels. It then tweens between the defined states and pauses at each state.

#### Arguments:
* `states`: The unquoted name of the column holding the state levels in the data.
* `transition_length`: The relative length of the transition. Will be recycled to match the number of states in the data
* `state_length`: The relative length of the pause at the states. Will be recycled to match the number of states in the data
* `wrap`: Should the animation wrap-around? If TRUE the last state will be transitioned into the first.

## _Racing_ Bar Graph {.smaller}
`ease_aes("cubic-in-out")`

#### Usage:

`ease_aes(default = "linear", ...)`

#### Description:
Easing defines how a value change to another during tweening. Will it progress linearly, or maybe start slowly and then build up momentum. In `gganimate`, each aesthetic or computed variable can be tweened with individual easing functions using the `ease_aes()` function. All easing functions implemented in tweenr are available, see `tweenr::display_ease` for an overview. Setting an ease for x and/or y will also affect the other related positional aesthetics (e.g. `xmin`, `yend`, etc).

#### Functions
* `cubic`: Models a power-of-3 function
* `-in-out`: The first half of the transition it is applied as-is, while in the last half it is reversed

## _Racing_ Bar Graph {.smaller}
Lastly, set up some aesthetics:
```{r racing bars animate aes, eval = FALSE}
# plot
p <- ordered_df %>% 
  ggplot(aes(ordering, group = topic)) +
  geom_col(aes(y = cum_total, width = 0.9, fill = topic)) +
  geom_text(aes(y = cum_total, label = topic), 
            hjust = "left", fontface = "bold", nudge_y = 50) +
  coord_cartesian(clip = "off", expand = FALSE) +
  scale_fill_viridis_d() +
  coord_flip() +
# animate
  transition_states(Year, transition_length = 8, state_length = 4, wrap = FALSE) +
  ease_aes("cubic-in-out") +
# aesthetics
### <b>
  labs(subtitle = "Trends in sequencing methods", 
       title = "Year {closest_state}", y = "cumulative total papers") +
  theme(plot.background = element_blank(), legend.position = "none",
        axis.ticks.y = element_blank(), axis.text.y = element_blank(),
        text = element_text(size=14), plot.title = element_text(size = 35)) +
  ylim(0,1300) +
  xlab("") 
### </b>
```

## _Racing_ Bar Graph
Let's animate!

The figure shown on the next slide was generated using these parameters:
```{r render animation high, eval = FALSE}
# render the animation
animate(p, nframes = 750, fps = 20, end_pause = 10)
```

However, due to the lengthy time it takes to generate, we should reduce these numbers:
```{r render animation, eval = FALSE}
# rendering the animation
animate(p, nframes = 100, fps = 5, end_pause = 10)
```


## _Racing_ Bar Graph {.smaller}
`animate(p, nframes = 100, fps = 5, end_pause = 10)`

#### Usage:
`animate(plot, ...)`

#### Description:
This function takes a `gganim` object and renders it into an animation. The nature of the animation is dependent on the renderer, but defaults to using `gifski` to render it to a gif. The length and framerate is decided on render time and can be any two combination of `nframes`, `fps`, and `duration`. Rendering is happening in discrete time units.

#### Arguments
* `plot`: A `gganim` object
* `nframes`: The number of frames to render (default `100`)
* `fps`: The framerate of the animation in frames/sec (default `10`)
* `duration`: The length of the animation in seconds (unset by default)
* `start_pause`,`end_pause`: Number of times to repeat the first and last frame in the animation (default is `0` for both)

## _Racing_ Bar Graph {.vcenter .flexbox}
```{r final animation, out.width="500px", echo = FALSE}
knitr::include_graphics("https://github.com/dy-lin/hs19-trends/raw/master/figures/sequencing_race.gif")
```
<center><font size="3">Source:https://dy-lin.github.io/hs19-trends/R/general_vis.html</font></center>

<!--## Resources
* How to Create a Bar Chart Race in R – Mapping United States City Population 1790-2010
    * https://www.r-bloggers.com/how-to-create-a-bar-chart-race-in-r-mapping-united-states-city-population-1790-2010/
* Bar Chart Races With `gganimate`
    * https://emilykuehler.github.io/bar-chart-race/-->
    
# Sankey Diagrams

## What is a Sankey Diagram? {.vcenter .flexbox .smaller}
* _Sankey Diagram_: a type of flow diagram where the width of the arrows is proportional to the flow rate
```{r sankey example, out.width="450px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://github.com/dy-lin/hs19-trends/raw/master/workshop/sankey_example.png")
```
<center><font size="3">Source: https://cran.r-project.org/web/packages/googleVis/vignettes/googleVis_examples.html</font></center>


## Sankey Diagrams
Here are the packages we need to load for a Sankey diagram:
```{r load lib, message = FALSE, warning = FALSE}
library(tidyverse)
library(ggrepel)
library(grid)
library(ggalluvial)
library(egg)
```
## Sankey Diagram {.smaller}
Let's load in our pre-processed dataset:
```{r load sankey data, message = FALSE, warning = FALSE}
base <- "https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/"
SK <- "data/sankey-processed.csv"
url <- paste0(base, SK)
datSK <- read_csv(url)
```
```{r print datSK, echo = FALSE}
datSK %>% head() %>% kable()
```

## Sankey Diagram {.smaller}
Here's an overview of plotting the Sankey diagram using `ggplot`:
```{r plot sankey, eval = FALSE}
sankey <- ggplot(datSK, aes(y = Weight, axis1 = From, axis2 = To)) +
  geom_alluvium(aes(fill = From), width = 1 / 12) +
  geom_stratum(alpha = 0, width = 1 / 12, color = "black") +
  scale_x_discrete(limits = c("From", "To"), expand = c(0.3, 0.1)) +
  scale_fill_viridis_d() +
  theme_void() +
  theme(axis.title.y = element_blank(), axis.title.x = element_blank(),
    axis.ticks.x = element_blank(), axis.ticks.y = element_blank(),
    axis.text.x = element_blank(),axis.text.y = element_blank(),
    legend.position = "none", plot.title = element_text(hjust = 0.5)) +
  ggrepel::geom_label_repel(
    aes(label = From), stat = "stratum", size = 3, direction = "x", hjust = 10) +
  ggrepel::geom_label_repel(
    aes(label = To), stat = "stratum", size = 3, direction = "y", nudge_x = 0.5) +
  geom_label(aes(label = Weight), stat = "stratum", alpha = 0.8) +
  ggtitle("Top 10 Institutions Publications By Topic")

sankey <- set_panel_size(sankey, width  = unit(18, "cm"), height = unit(10, "cm"))
grid.newpage()
grid.draw(sankey)
```

## Sankey Diagram {.smaller}
Let's take a closer look at the `geom` functions used:
```{r plot sankey geom, eval = FALSE}
sankey <- ggplot(datSK, aes(y = Weight, axis1 = From, axis2 = To)) +
### <b>
  geom_alluvium(aes(fill = From), width = 1 / 12) +
  geom_stratum(alpha = 0, width = 1 / 12, color = "black") +
### </b>
  scale_x_discrete(limits = c("From", "To"), expand = c(0.3, 0.1)) +
  scale_fill_viridis_d() +
  theme_void() +
  theme(axis.title.y = element_blank(), axis.title.x = element_blank(),
    axis.ticks.x = element_blank(), axis.ticks.y = element_blank(),
    axis.text.x = element_blank(),axis.text.y = element_blank(),
    legend.position = "none", plot.title = element_text(hjust = 0.5)) +
  ggrepel::geom_label_repel(
    aes(label = From), stat = "stratum", size = 3, direction = "x", hjust = 10) +
  ggrepel::geom_label_repel(
    aes(label = To), stat = "stratum", size = 3, direction = "y", nudge_x = 0.5) +
  geom_label(aes(label = Weight), stat = "stratum", alpha = 0.8) +
  ggtitle("Top 10 Institutions Publications By Topic")

sankey <- set_panel_size(sankey, width  = unit(18, "cm"), height = unit(10, "cm"))
grid.newpage()
grid.draw(sankey)
```

## Sankey Diagram {.smaller}
`geom_alluvium(aes(fill = From), width = 1 / 12)`

#### Usage:
`geom_alluvium(mapping = NULL, data = NULL, stat = "alluvium", position = "identity", width = 1/3, knot.pos = 1/6, na.rm = FALSE, show.legend = NA, inherit.aes = TRUE, ...)`

#### Description:
`geom_alluvium` receives a dataset of the horizontal (`x`) and vertical (`y`, `ymin`, `ymax`) positions of the __lodes__ of an alluvial diagram, the intersections of the alluvia with the strata. It plots both the lodes themselves, using `geom_lode()`, and the flows between them, using `geom_flow()`.

#### Arguments:
* `mapping`: Set of aesthetic mappings created by `aes()`
* `width`: Numeric; the width of each stratum, as a proportion of the distance between axes. Defaults to 1/3.

## Sankey Diagram {.smaller}
`geom_stratum(alpha = 0, width = 1 / 12, color = "black")`

#### Usage:
`geom_stratum(mapping = NULL, data = NULL, stat = "stratum", position = "identity", show.legend = NA, inherit.aes = TRUE, width = 1/3, na.rm = FALSE, ...)`

#### Description:
`geom_stratum` receives a dataset of the horizontal (`x`) and vertical (`y`, `ymin`, `ymax`) positions of the strata of an alluvial diagram. It plots rectangles for these strata of a provided `width`.

#### Arguments:
* `width`: Numeric; the width of each stratum, as a proportion of the distance between axes. Defaults to 1/3.

## Sankey Diagram

```{r final sankey, out.width="800px", echo = FALSE}
knitr::include_graphics("https://dy-lin.github.io/hs19-trends/R/Sankey.png")
```
<center><font size="3">Source: https://dy-lin.github.io/hs19-trends/R/visualization_sankey.html</font></center>

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width=12, fig.height=8)
```

# Mapping
## Mapping 
We want to look at several trends 

- Most Common topic per country 
- Number of papers per country 

## Mapping
```{r, out.width="700px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://dy-lin.github.io/hs19-trends/workshop/Map_example.png")
```

## Mapping 
Here are all the packages we need to format the data and do the plotting 

```{r load lib emma, message = FALSE, warning = FALSE}
library(ggplot2)
library(dplyr)
library(plotly)
library(here)
library(ggmap)
library(viridis)
library(rgeos)
library(maptools)
library(maps)
library(sf)
library(readr)
library(tidyverse)
library(knitr)
library(broom)
```
## Mapping 
Packages needed for plotting 

`ggplot2`: package for data visualizations 

`plotly`: makes interactive visualizations

`viridis`: good colour palatte for dealing with colourblindness and greyscale issues 

## Mapping 
There is a lot of code in the worksheet that describes how we got the data into the format we needed

-We will do one global plot and one of Europe 

## Mapping {.smaller}
Here is an example of what the data looks like:

```{r,include=FALSE}
world_data2=read.csv("https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/data/world_data2.csv")
```

```{r print overview table emma, echo = FALSE}
world_data2[14572:14580,2:9] %>% kable()
```

## Mapping 

```{r,message = FALSE, warning = FALSE,eval=FALSE}
pl <- ggplot() +
  ### <b>
  geom_polygon(data = world_data2, aes(x = long, y = lat,
  group = group,fill = log(Count),text=Count)) +
  ### </b>
  coord_fixed(1.3)+
  scale_fill_viridis()+
  theme_void()

world_plotly=ggplotly(pl,tooltip = "text")
```
## Mapping

```{r,message = FALSE, warning = FALSE,eval=FALSE}
pl <- ggplot() +
  geom_polygon(data = world_data2, aes(x = long, y = lat,
  group = group,fill = log(Count),text=Count)) +
  ### <b>
  coord_fixed(1.3)+
  scale_fill_viridis()+
  theme_void()
### </b>

world_plotly=ggplotly(pl,tooltip = "text")
```

## Mapping 
```{r, out.width="700px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://dy-lin.github.io/hs19-trends/workshop/world_map_example.png")
```

## Mapping 
```{r,include=FALSE}
europe_data=read.csv("https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/data/europe_data.csv")
points_europe=read.csv("https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/data/europe_points.csv")
```
If we want to look at certain countries we can do that too 

- depending on how much data is in one region this may be a good idea 
- will have to figure out the coordinates you want 
- plot will be set up similarily to the global one but we will add in new elements 

## Mapping{.smaller}

```{r,echo=FALSE}
europe_data[67:72,]%>%kable()
```

## Mapping 

```{r,message = FALSE, warning = FALSE,eval=FALSE}
## <b>
 pl <- ggplot() + 
  geom_polygon(data = europe_data, aes(x = long, y = lat, 
  group = group, fill = log(Count),
  text=paste(region,Count, sep=";"))) +
  ## </b>
  geom_point(data=points_europe,aes(x=X,y=Y,
  text=str_wrap(affiliation,50)),
  alpha=0.5,size=0.5,colour="grey")+
  coord_fixed(1.3)+
  scale_fill_viridis()+
  theme_void()
```

## Mapping 
```{r,message = FALSE, warning = FALSE,eval=FALSE}
 pl <- ggplot() + 
  geom_polygon(data = europe_data, aes(x = long, y = lat, 
  group = group, fill = log(Count),
  text=paste(region,Count, sep=";"))) +
  ## <b>
  geom_point(data=points_europe,aes(x=X,y=Y,
  text=str_wrap(affiliation,50)),
  alpha=0.5,size=0.5,colour="grey")+
  ## </b>
  coord_fixed(1.3)+
  scale_fill_viridis()+
  theme_void()
```

## Mapping 

```{r, out.width="700px", echo = FALSE, fig.align = 'center'}
knitr::include_graphics("https://dy-lin.github.io/hs19-trends/workshop/europe_map_example.png")
```


# Bigrams 

## Bigrams 

__Bigrams__ are two consecutive words within a given text input 

  - ex. We love coding in R
  - (We Love), (Love coding), (coding in), (in R)
  
## Bigrams 
```{r, message = FALSE, warning = FALSE}
library(tidyverse)
library(tidytext)
library(tm)
library(widyr)
library(igraph)
library(ggplot2)
library(ggraph)
library(readr)
library(tidygraph)
```

## Bigrams 

`ggraph`: visualizations for network structures 

`ggplot2`: package for data visualizations 

## Bigrams

 __visualize_bigrams__: extracts bigrams from a text field, calculates frequency of bigrams, and creates a bigram plot to visualize relationships between words
    
* `df_name`: name of dataframe that contains the text field of interest
* `textfield`: name of text field (ie. column name)
    
```{r,eval=FALSE}
visualize_bigrams <- function(df_name, textfield, topic_title)
```

## Bigrams 
In the worksheet, you will see a lot of code that just talks about the formatting of the data. If you are interested in that look into it, but we're just covering the plotting 
```{r,include=FALSE}
visualize_bigrams <- function(df_name, textfield, topic_title){
  
    # Create frequencies of bigrams
    df_cleaned <- df_name %>% 
      mutate(textfield_clean = removeWords(gsub("[^A-Za-z0-9 ]", "", {{textfield}}), stop_words$word))
    
    df_bigrams <- df_cleaned %>%
      unnest_tokens(bigrams, textfield_clean, token = "ngrams", n = 2)
    
    df_freq <- as.data.frame(table(df_bigrams$bigrams)) %>% 
      arrange(desc(Freq))
    
    # Visualizations
    df_top_bigrams <- df_freq %>%
      top_n(100, Freq) %>% 
      separate(Var1, c("word1", "word2"))
    
    top_bigram_words <- c(df_top_bigrams$word1, df_top_bigrams$word2) %>%
      unique()
    
    word_list <- df_cleaned %>%
      unnest_tokens(words, textfield_clean, token = "ngrams", n = 1) 
    
    df_word_list <- as.data.frame(table(word_list$words)) %>% 
      arrange(desc(Freq)) %>%
      filter(Var1 %in% top_bigram_words)
    
    names(df_word_list)[2] <- "Term_Frequency"
    names(df_top_bigrams)[3] <- "Edge_Frequency"
    
    graph_from_data_frame(vertices =  df_word_list, d = df_top_bigrams) -> graph_hold
    
    pl <- graph_hold %>%
      ggraph(layout = "fr") +
      geom_edge_link(aes(edge_colour = log(Edge_Frequency)), show.legend = TRUE) +
      geom_node_point(aes(color = Term_Frequency, size = Term_Frequency), alpha = 0.7) +
      scale_fill_viridis_c() +
      scale_edge_color_viridis(direction=-1)+
      geom_node_text(aes(label = name), repel = TRUE) +
      scale_color_viridis_c(direction = -1) +
      theme_void() +
      guides(size=FALSE) +
      labs(title = quo_name(topic_title)) +
      theme(plot.title = element_text(size = 26, face = "bold"))
    
    ggsave(pl,filename = paste0("../figures/", "bigrams_", str_to_lower(str_replace(topic_title, "\\s", "_")), ".png")
           ,width = 12
           ,height = 8)
    pl
}
```

## Bigrams 
```{r,message = FALSE, eval=FALSE}
pl <- graph_hold %>%
      ggraph(layout = "fr") +
      geom_edge_link(aes(edge_alpha = Edge_Frequency), 
      show.legend = TRUE) +
      geom_node_point(aes(color = Term_Frequency, 
      size = Term_Frequency), alpha = 0.7) +
      scale_fill_viridis_c() +
      geom_node_text(aes(label = name), repel = TRUE) +
      scale_color_viridis_c(direction = -1) +
      theme_void() +
      guides(size=FALSE) +
      labs(title = quo_name(topic_title)) +
      theme(plot.title = element_text(size = 26, face = "bold"))
```

## Bigrams {.vcenter .flexbox}
```{r, include=FALSE}
df=read.csv("https://raw.githubusercontent.com/dy-lin/hs19-trends/master/workshop/data/bigrams.csv")
```

``` {r,echo=FALSE}
print(kable(unique(df$topic),col.names = "Search Term"))
```

## Bigrams 
```{r,message = FALSE,warning=FALSE,eval=FALSE}
df_assembly <- df %>% 
  filter(topic == "Assembly")
visualize_bigrams(df_assembly, abstract, "")
```


## Bigrams
```{r bigrams final, out.width="800px", echo = FALSE}
knitr::include_graphics("https://dy-lin.github.io/hs19-trends/figures/bigrams_assembly.png")
```



## GitHub
For more information regarding the Hackseq project:
https://github.com/dy-lin/hs19-trends